docs/source/cpp_tutorial.rst

```diff
@@ -0,0 +1,448 @@
+========================
+C++ Walkthrough Tutorial
+========================
+
+Table of Contents
+=================
+
+- `Step 1: Starting off with cuVS`_
+
+- `Step 2: Generate some data`_
+
+- `Step 3: Using brute-force indexes`_
+
+- `Step 4: Using the ANN indexes`_
+
+- `Step 5: Evaluate neighborhood quality`_
+
+- `Advanced Features`_ 
+
+  * `Serialization`_
+  
+  * `Filtering`_
+  
+  * `Stream Pools`_
+  
+  * `Device Resources Manager`_
+  
+  * `Device Memory Resources`_
+  
+  * `Workspace Memory Resource`_
+
+cuVS has several important algorithms for performing vector search on the GPU and this tutorial walks through the primary vector search APIs from start to finish to provide a reference for quick setup and C++ API usage.
+
+This tutorial assumes cuVS has been installed and/or added to your build so that you are able to compile and run RAFT code. If not done already, please follow the [build and install instructions](build.md) and consider taking a look at the [example c++ template project](https://github.com/rapidsai/raft/tree/HEAD/cpp/template) for ready-to-go examples that you can immediately build and start playing with. Also take a look at RAFT's library of [reproducible vector search benchmarks](raft_ann_benchmarks.md) to run benchmarks that compare cuVS against other state-of-the-art nearest neighbors algorithms at scale.
+
+For more information about the various APIs demonstrated in this tutorial, along with comprehensive usage examples of all the APIs offered by RAFT, please refer to the `cuVS C++ API Documentation <>https://docs.rapids.ai/api/cuvs/nightly/cpp_api/>`_.
+
+Step 1: Starting off with cuVS
+==============================
+
+CUDA Development?
+-----------------
+
+If you are reading this tuturial then you probably know about CUDA and its relationship to general-purpose GPU computing (GPGPU). You probably also know about Nvidia GPUs but might not necessarily be familiar with the programming model nor GPU computing. The good news is that extensive knowledge of CUDA and GPUs are not needed in order to get started with or build applications with RAFT. RAFT hides away most of the complexities behind simple single-threaded stateless functions that are inherently asynchronous, meaning the result of a computation isn't necessarily read to be used when the function executes and control is given back to the user. The functions are, however, allowed to be chained together in a sequence of calls that don't need to wait for subsequent computations to complete in order to continue execution. In fact, the only time you need to wait for the computation to complete is when you are ready to use the result.
```

Comment:
This feels like how to get started with RAFT to build a GPU algorithm, not how to get started with cuVS to do vector search?

Reply:
No, this is eventually going to be a guide on how to use cuVS APIs end-to-end. It's not being linked into the docs yet so it's not even reachable atm. Ran out of time.

Reply:
I don't want to remove it only because it took time to convert it from markdown to rust, but I'm hoping to revise it and link it in sometime during 24.12 (as time permits).